### ISSUE TITLE
Coverage reports?

### ISSUE THREAD

**Issue by willvincent** [bitwes/Gut#81]
I've recently started using istanbul for coverage report generation of nodejs code and boy it sure is helpful, I can see down to the individual line whether code is covered by tests or not.

Any ideas how one might go about generating test coverage reports? Would be a nice feature.  

I suspect it would most likely happen in a separate package, but I'm not sure how one would go about writing such a tool from scratch.

**Comment by bitwes** [comment]
I've thought about this as well and I'm also at a loss of how they do this.  Seems like a lot of magic.  It might be easier as an engine plugin, written in c or c++, whichever the engine is written in.  I think leveraging the existing parser would be helpful.  

I agree it would be great to have but it's not going n my short list.  I'll leave this issue open for awhile to see if anyone else has any ideas or wants to tackle it.

**Comment by willvincent** [comment]
Yeah it seems like it must have to involve some lower level access to the interpreter to know what files are being loaded and so forth.  Maybe I'll poke around at the source of istanbul and try to gleen some insight into how they manage it with node.

**Comment by willvincent** [comment]
Ran across this.. http://www.semdesigns.com/Company/Publications/TestCoverage.pdf

Pretty high level, but at least explains the necessary process.

Basically:
- Parse the code to identify the basic blocks/generate [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)
- Traverse the syntax tree and rewrite code to include markers that set a flag when that syntax block is accessed
- Run tests against instrumented code
- Compile report by reviewing which parts of the tree were actually visited

In practice that's not going to be _simple_ ... but could probably be baked in to the gdscript interpreter so that it automatically instruments and returns a report if run with a special flag or something.

**Comment by bitwes** [comment]
That doc was a good find, thanks for that.

This does sound feasible.  I think the hard part will be wiring it in.  I have no idea what it looks like to plugin to the gdscript interpreter but it sounds kinda fun.

It could be done in gdscript if we have the user do it manually by creating instances of "covered" classes similar to how the doubles work now.  You could then get a feel for what was covered in your class but you wouldn't get any visibility of anything you didn't specifically create as "covered".

If we could hijack `load` and `preload` somehow...for EVERYTHING...then we could do this...I don't know of any way to ["monkey patch"](https://en.wikipedia.org/wiki/Monkey_patch) in gdscript.

I'll keep mulling this one over.  It pops into my head from time to time and this gives me more to think about.  Thanks.

**Comment by VictorRogers** [comment]
I just wanted to cast a vote for this feature idea. This is pretty much the one thing I'm missing while developing for Godot.

**Comment by bitwes** [comment]
Given the world has shut down, I started working extra hard on the one game I've been making.  No sense making a testing tool if I don't ever make a game.  Haha.  So the answer is that I haven't had anytime to revisit this but I don't have any plans to close this issue.  It'll be a great feature when I get time to implement another big feature for GUT.

I would be more than happy to help anyone out if they were interested in trying.  I don't have time to do it right now, but I'd make time to have to do less of the work.

**Comment by felix-hellman** [comment]
I decided to write a proof of concept implementation of injecting instrumented code in gdscript separate from GUT. It seems to solve all the major "problems" however I still need to figure out how to integrate this with GUT and do some general cleanup of the code, however I think I can do this.

Since hijacking load and/or preload was out of the question I went with connecting to "on_tree_enter" signal. So it only works when "add_child" is called. I.e. Orphans will not get coverage.

The inject_test_metrics(node) method injects checkpoints in the form of signals that gets sent to a TestSuite.

I just want to make sure that this would be an OK approach to the problem before continuing with the integration towards GUT.

Pasting some example code of how it could be used along with example output.
Line coverage calculation needs to be tweaked a bit, but it kind of works.
### Example class to be tested:

```gdscript
extends Node2D


class_name Other, "res://scripts/Other.gd"

func _ready():
	pass

func test():
	print("hi")
	
func conditional(condition):
	if condition:
		print("Hi")
	else:
		print("Bye")
```
### Example of measuring coverage
```gdscript
extends Node2D

var injector : TestMetricInjector

func _ready():
	get_tree().connect("node_added", self, "on_tree_enter")
	injector = load("res://TestMetricInjector.tscn").instance()
	for script in load_scripts(): # Discovery of scripts to add coverage for
		var instance = load(script).new()
		add_child(instance)
	
	
	print(injector.get_test_report())
	one()
	print(injector.get_test_report())
	two()
	print(injector.get_test_report())
	three_a()
	print(injector.get_test_report())
	three_b()
	print(injector.get_test_report())

func on_tree_enter(node):
	node = injector.inject_test_metrics(node)
	
func one():
	var o : Other = Other.new()
	add_child(o)
	o.test()
	
func two():
	var o : Other = Other.new()
	add_child(o)
	o._ready()
	
func three_a():
	var o : Other = Other.new()
	add_child(o)
	o.conditional(false)

func three_b():
	var o : Other = Other.new()
	add_child(o)
	o.conditional(true)
	
func load_scripts():
	var dir : Directory = Directory.new()
	var base_path = "res://scripts"
	var scripts = []
	if dir.open(base_path) == OK:
		dir.list_dir_begin()
		var file_name = dir.get_next()
		while file_name != "":
			if not dir.current_is_dir() and ".gd" in file_name:
				var path = base_path + "/" + file_name
				scripts.append(path)
			file_name = dir.get_next()
	else:
		print("An error occurred when trying to access the path.")
	return scripts
```
### Output from example:
```
{res://scripts/Other.gd:{line_coverage:14.285714%, method_coverage:33.333333%, uncovered_lines:[9, 10, 11, 12, 13, 13, 14, 15, 15, 16]}}
hi
{res://scripts/Other.gd:{line_coverage:42.857143%, method_coverage:66.666667%, uncovered_lines:[12, 13, 13, 14, 15, 15, 16]}}
{res://scripts/Other.gd:{line_coverage:42.857143%, method_coverage:66.666667%, uncovered_lines:[12, 13, 13, 14, 15, 15, 16]}}
Bye
{res://scripts/Other.gd:{line_coverage:71.428571%, method_coverage:100%, uncovered_lines:[13, 14, 15]}}
Hi
{res://scripts/Other.gd:{line_coverage:100%, method_coverage:100%, uncovered_lines:[]}}
```

**Comment by bitwes** [comment]
This is very promising!  Well done.  I think this approach will cover most of the use cases.  I've got some questions but this is __really really__ exciting to see.

1.  What does `injector.inject_test_metrics(node)` return?  My thought was that it was doing something similar to what `double` in GUT does (dynamically making a new object) but then what is the value of `o` is in each line of the `one()` method for example?  Is `func on_tree_enter(node):` actually changing what `node` points to?
```
func on_tree_enter(node):
	node = injector.inject_test_metrics(node)
	
func one():
	var o : Other = Other.new()
	add_child(o)
	o.test() # is this pointing to something different than the instance created 2 lines up?
```

2.  Is there a way to get coverage of a class that extends `Reference` and is not added to the tree?  Is it just a matter of calling `injector.inject_test_metrics(node)` on the instance or class?  

3.  How does the "discovery" logic you posted factor into the injection?

**Comment by felix-hellman** [comment]
Sweet!

1. It does return the same object that went in however it is mutated. So actually we don't need to assign node again, since this basically does nothing. It's enough with the following.
```
func on_tree_enter(node):
	injector.inject_test_metrics(node)
```
2. Yes, you just need to call `injector.inject_test_metrics(node)` for a class that extends Reference

3. Not much, it's basically just making sure all scripts get loaded so even scripts with 0% coverage gets "picked up", I just hastily put this together to see if it works.

I will take a look at how to integrate this into GUT then. Any pointers on where it should go will be appreciated.

**Comment by bitwes** [comment]
How are you mutating the code of the object?  As far as I knew there wasn't a way to dynamically change the implementation of a method at runtime?  You can change the script associated with an object at runtime, but not the implementation of the script that is currently loaded.

As for integrating with GUT.  There's a lot of unknowns at this point.  If you can make an example test script that uses your `Injector` to test something then we can get a better feel on what all will need to be added.  Maybe we could Beta this feature as an addon to GUT or something that is a little more manual to use at first.

**Comment by felix-hellman** [comment]
You are correct, but you can circumvent this with a bit of a hack. Here is a minimal example
```
func _ready()
	var other : Other = Other.new()
	add_child(other)
	modify_impl(other)
	other.new_method()
	
func modify_impl(obj):
	var script : Script = GDScript.new()
	var source = obj.get_script().get_source_code()
	script.set_source_code(source + "\n\nfunc new_method():\n\tprint(\"New Method!\")\n")
	script.reload()
	obj.set_script(script)
```

Sure thing. I will fork GUT and see if I can use it in some test case. Maybe I can add it as a opt-in flag which defaults to off for now.

**Comment by bitwes** [comment]
Ok, yea, that's what `addons/gut/doubler.gd` does.  

This feels like it's going to behave very similar to the way the doubler and stubber do in GUT.  `doubler.gd` is responsible for creating dynamic classes/scenes that can be instanced.  These altered instances then call back to a class that tracks interactions.

If you haven't already, take a look at:
* https://github.com/bitwes/Gut/blob/master/addons/gut/doubler.gd
  * Note that the doubler used to make files but now does everything dynamically.  There's still some cruft associated with the old approach that can be misleading.
* https://github.com/bitwes/Gut/blob/master/addons/gut/stubber.gd
* https://github.com/bitwes/Gut/blob/master/addons/gut/spy.gd
* https://github.com/bitwes/Gut/blob/master/addons/gut/double_templates/script_template.txt

Then trace through `gut.gd` watching how it interacts with its instances of the stubber and spy (`_doubler`, `_spy`).  Also watch how `test.gd` interacts with the doubler (mainly through the `_smart_double` method).

My approach would be to get something like the following to work.  After that I'd start making it more user friendly and integrating it with GUT.
```
# test_foo.gd
extends "res://addons/gut/test.gd"

var TestMetricInjector = load(<path to it>).new()
var Foo = load("res://scripts/foo.gd")
# This returns a loaded class and not an instance, instancing will be done in 
# each test or in before_all or before_each
var CoveredFoo = TestMetricInjector.create_covered(Foo)

func after_all():
  TestMetricInjector.print_report()

func test_foo_thing():
  var f = CoveredFoo.new()
  # do stuff
  assert_true(f.something())

# more tests
```

Really looking forward to what you come up with.  Don't hesitate to ask any questions.

**Comment by felix-hellman** [comment]
I have gotten the following to work in GUT so far.

- Enable test coverage via a flag in the plugin_control.
- Listen to signal for test end to log test-report when done
And this basic example
```
extends 'res://addons/gut/test.gd'

var loaded_dummy = load("res://scripts/dummy.gd")
var dummy

func before_each():
	dummy = loaded_dummy.new()
	add_child(dummy) #If coverage flag is active we listen to the on_tree_enter signal
	

func test_test():
	dummy.conditional(true)

func after_all():
	for child in get_children():
		remove_child(child)
```

I can also run most existing tests (although it does not automagically get you any test-coverage but it does not crash either), however it seems to be a problem with tests that does not clean up their children after a test (and coverage is enabled) causes some kind of error.

The easiest way to reproduce this is to remove the implementation of `after_all` in the test above.

The error is thrown in `_get_obj_filename`

![Error](https://i.imgur.com/S8Ja07R.png)

This error does not get thrown when i disable the flag for test-coverage. 
So I'm pretty sure it's my fault.

Since it works somewhat I'll take a look at this error later, for now I'll try to clean up the integration a bit and actually follow the coding standard of the project.

When it works, it looks like this. 
![Coverage](https://i.imgur.com/oHiPGlp.png)

I also tried generating reports for bigger files and it can get really cluttered when logging all uncovered_lines. But I don't have any idea of how to present this in a better way right now. Any suggestions are welcome.

Things to do:

1. Cleanup integration
2. Better way to present data
3. Fix error

**Comment by hootener** [comment]
> But I don't have any idea of how to present this in a better way right now. 

I can maybe help here a bit. Most coverage reporting tools (to my knowledge) don't actually output much of anything to the terminal. It's typically a two step approach:

1. output some sort of machine readable format (.xml, .json, etc)
2. Use a secondary process to turn that machine readable format into something human readable and present it as html so the user can open it in their browser. 

Regarding step 1, there are ton of popular formats that I recommend looking into and just copying their approach. There's no need to reinvent the wheel here. Some popular formats:

1. Cobertura (seen with python and other langauge)
2. Clover (usually in the .NET and C# space)
3. JSON  (ruby/rspec)

I make no recommendation that you have to use a particular format, but I do recommend adhering to _some_ standard so that GUT coverage reports will play nice with third party services (e.g., Codecov, Coveralls, Code Climate, etc). If you'd like some inspiration, here's an example of a json output format:

```
{
  "coverage": {
    "filename": {
      "1": 0,      # line 1 missed
      "2": 1,      # line 2 hit once
      "3": "1/2",  # line 3 partial hit, one missing branch
      "4": "1/3",  # line 4 partial hit, two missing branches
                             # skip line 5
      "6": null,   # skip line 6
      "7": 5       # line 7 hit 5 times
    }
  }
}
```

Here's cobertura for a single file, named adder.py:

```
<?xml version="1.0" ?>
<coverage version="5.2.1" timestamp="1596599468873" lines-valid="27" lines-covered="12" line-rate="0.4444" branches-covered="0" branches-valid="0" branch-rate="0" complexity="0">
	<!-- Generated by coverage.py: https://coverage.readthedocs.io -->
	<!-- Based on https://raw.githubusercontent.com/cobertura/web/master/htdocs/xml/coverage-04.dtd -->
	<sources>
		<source>/home/hootener/Source/codecov/codecov-assume-flag-test/src</source>
	</sources>
	<packages>
		<package name="adder" line-rate="0.8333" branch-rate="0" complexity="0">
			<classes>
				<class name="__init__.py" filename="adder/__init__.py" complexity="0" line-rate="1" branch-rate="0">
					<methods/>
					<lines/>
				</class>
				<class name="adder.py" filename="adder/adder.py" complexity="0" line-rate="0.8333" branch-rate="0">
					<methods/>
					<lines>
						<line number="1" hits="1"/>
						<line number="3" hits="1"/>
						<line number="4" hits="1"/>
						<line number="8" hits="1"/>
						<line number="10" hits="1"/>
						<line number="11" hits="0"/>
					</lines>
				</class>
			</classes>
		</package>
	</packages>
</coverage>

```

**Comment by hootener** [comment]
Oh! Another point I wanted to make:

> I also tried generating reports for bigger files and it can get really cluttered when logging all uncovered_lines. 

This is okay, really. Coverage reports can get enormous, which is one of the main reasons I think the results aren't piped to the terminal after a test run completes. This is another reason that you typically see coverage output from a machine readable file translated to HTML and presented that way. Here's an example from coverage.py, which processes cobertura and output an html bundle" https://nedbatchelder.com/files/sample_coverage_html/index.html

I echo @bitwes 's sentiments from earlier in the thread, this progress is _really_ exciting to see, and I hope it makes it to release. If so, I look forward to supporting Godot Unit Test coverage in Codecov! :)

**Comment by dodgyville** [comment]
I would love to use code coverage as a quality gate on my game repo. If it was output as an intermediate file format it can be parsed for a variety of uses.

**Comment by koalafr** [comment]
Hello, I have implemented a stupid bash tool I use to get the coverage report badge on gitlab

**I have no significant experience in unit testing and TDD** so it's more of a _guide_ for me than a full featured tool.
This tool ignores the file contents. Maybe the tool can be extended with real coverage data, or include matching files in xml.

It's capable of printing some infos, a percentage and exporting a fake xml cobertura report
I kept it separate from GUT / GODOT so it would be hopefully more flexible/adaptable to updates, and I like BASH also

This is the solution I am currently using and it integrates well with Gut, so big thank you @bitwes 

I've released the tool with a yaml example under MIT here : https://github.com/koalafr/godot-coverage-hack/

Any input welcome

eg output

```
Running Godot coverage hack by kbs...

Looking for gdscript files...
2 gdscript files found

[✓] Done indexing scripts

Looking for unit test files...
2 unit test files found

[✓] Done indexing tests

Running basic code (file) coverage...

[✓] Found unit test for human.gd
[X] Missing test for koala.gd

[!] Warning: test without matching script: test_cat.gd

--- Coverage Report ---

- Scripts -
Tested:          1
Untested:        1
Total:           2

- Tests -
Valid:           1
Unused:          1
Total:           2

Coverage:        50%

[✓] Coverage report finished
```

**Comment by felix-hellman** [comment]
@bitwes @hootener 

I got stuck on some issue last time and kind of forgot about this, sorry. Turns out the other implementation was a bit hacky and naive which made branching an issue.
I now have something that can generate reports over here, https://github.com/felix-hellman/godot-code-coverage

It is written as a package which I will add to the package manager that I am maintaining.

Not really ready for prime time just yet, but the hard things are solved.

I'll make a new attempt at integrating it into GUT with a feature toggle if it's still needed.

Reports follow this format (.json)
```json
{
  "coverage": {
    "res://pkg/felix-hellman-codecoverage/test/CodeToCover.gd": {
      "t": {
        "8": "1/1",
        "9": "1/1",
        "10": "1/1",
        "11": "1/2",
        "12": "1/1",
        "13": "1/1",
        "15": "0/1",
        "16": "0/1",
        "17": "1/1",
        "19": "1/1",
        "20": "1/1",
        "21": "1/1"
      },
      "t2": {
        "24": "0/1",
        "25": "0/1",
        "26": "0/1",
        "27": "0/3",
        "28": "0/1",
        "29": "0/1",
        "30": "0/1",
        "31": "0/1",
        "32": "0/1",
        "33": "0/1",
        "34": "0/1",
        "36": "0/1",
        "37": "0/1",
        "38": "0/1"
      },
      "t3": {
        "41": "0/1",
        "42": "0/1",
        "43": "0/1",
        "44": "0/2",
        "45": "0/1",
        "47": "0/1",
        "48": "0/1"
      }
    }
  }
}
```
For this example
```
extends Node

class_name CTC, "res://pkg/felix-hellman-codecoverage/test/CodeToCover.gd"

var x = 123

func t(conditional : bool):
	var a = 1 + 1
	var b = 2 + 5
	a = b + 7
	if conditional:
		b = a - 14
		a = a * 2
	else:
		a = b - 14
		b = b * 2
	var c = a + b
	yield()
	a = a - a
	b = b - b
	return c

...
```
Result from this test
```
var to_cover = load("res://pkg/felix-hellman-codecoverage/test/CodeToCover.gd").new()
factory.inject_object(to_cover)
var f = to_cover.t(true)
f.resume()
factory.on_complete()
```

**Comment by MathisMartin31** [comment]
Hello @bitwes @felix-hellman !

I know this is an old thread, but I have stumbled on it while doing some research on unit testing and possible code coverage in a Godot project I'm preparing to show to my manager at work. I'm currently on Godot 4, and I took the liberty to clone Mr. Hellman's repo in my project and tinker with it.

What I did is certainly not perfect, but I got it working with Godot 4.4.stable and GUT 9.4.0 by doing this (I'm currently bypassing the AutoInject script entirely and using the TestReportFactory directly, but I saw you can add start and end Hooks for GUT to run before and after anything else so I'll try that next) :

```
extends GutTest

var report_factory := TestReportFactory.new()

func before_all():
	report_factory.inject_object(SomeSingleton)

func after_all():
	report_factory.on_complete()

func test_foo():
    SomeSingleton.foo()

func test_bar():
    SomeSingleton.bar()
```

After tests for SomeSingleton are done I am left with a coverage JSON that looks very much like the one in [this earlier message](https://github.com/bitwes/Gut/issues/81#issuecomment-1133940216). I don't know yet if it's 100% accurate because I changed some stuff in how things work, but it looks like it's really picking up lines visited by my tests.

Note that for now I am doing this by hand and I only tested what I've done on a singleton. I still need to try it with more complex nodes (with children and more a complex state), but I feel that having it run on the latest stable Godot4 is at least something to go on.

If someone is still interested in this I'll try opening a merge request in Mr. Hellman's repo (or maybe in here if that's more appropriate), so please let me know what you think 😃

**Comment by bitwes** [comment]
This is a VERY old thread, but I'd still be happy to see this as a feature.  I don't have the time to really dig into this though.  If you are going to try this out, it might be good to open up another issue where we can talk specifics.  I would like to see some psuedo code on how it might be used.

Could this be something that lives outside of GUT?

**Comment by MathisMartin31** [comment]
Yes now that I think about it this is totally something that could live outside of GUT as a separate addon. A few things that have popped up though:

- I have a few parsing quirks to iron out. Some are due to "weird" formatting practices but I need to take them into account.
- I didn't think of looking at it but I don't know if there might be a native way of generating the code's AST because that's the most difficult part. From there it's smooth sailing.
- As of now, the feature takes the form of an Autoload loading first that injects instrumentation into the GDScript code of nodes entering the tree, but I haven't found a way of adding such an autoload at runtime so I don't have to delete it when I export the game. Maybe making it a full-on addon or integrating to GUT might help.
- Lastly, the biggest limitation of this is that I can't cover static methods from classes. This works by taking the script of an object, adding signals and setting the new source code but for now I can't do that. I could instantiate the class and simply delete all the "static" keywords but there's obviously things that will go wrong with this approach.

**Comment by bitwes** [comment]
Found this today.  https://github.com/jamie-pate/godot-code-coverage/tree/main
